hail/src/main/resources/include/hail/NDArray.h

```diff
@@ -124,4 +124,40 @@ std::vector<long> unify_shapes(std::vector<long> &left, std::vector<long> &right
   return result;
 }
 
+// Assumption: left and right are of dimension 1 or more
+std::vector<long> matmul_shape(std::vector<long> &left, std::vector<long> &right) {
+  std::vector<long> result;
+  int l_size = left.size();
+  int r_size = right.size();
+
+  int left_inner_dim = l_size - 1;
+  int right_inner_dim = std::max(0, r_size - 2);
+  if (left[left_inner_dim] != right[right_inner_dim]) {
+    throw new FatalError("Invalid shapes for matrix multiplication");
+  }
+
+  if (l_size == 1 && r_size == 1) {
+    return result;
+  } else if (l_size == 1) {
+    result.assign(right.begin(), right.begin() + right_inner_dim);
+    result.push_back(right[r_size - 1]);
+  } else if (r_size == 1) {
+    result.assign(left.begin(), left.begin() + left_inner_dim);
+  } else {
+    assert(l_size == r_size);
+
+    for (int i = 0; i < l_size - 2; i++) {
```

Comment:
the rest of the code base prefers `++i`, which doesn't really matter with a modern C++ compiler, but let's be cool kids who don't unnecessarily ask for intermediate values to be kept around.